### PR TITLE
Handle k8s service events and trigger SE generation

### PR DESCRIPTION
**Pull request by aattuluri**
Changes include:
- Trigger SE generation for k8s service updates
- Order services by creation time to be more deterministic for multiple service matches
- Introduce additional delay during bootup before firing sync events for different types
- Reorder controller creation to start with services, gtps etc and then the deployment and rollouts
- Disable periodic sync for types other than deployment and rollout

### CODE REVIEW

File: admiral/pkg/clusters/registry.go [hunk 1]
```diff
@@ -120,31 +120,21 @@ func (r *RemoteRegistry) createCacheController(clientConfig *rest.Config, cluste
...
 
-	log.Infof("starting global traffic policy controller custerID: %v", clusterID)
-
-	rc.GlobalTraffic, err = admiral.NewGlobalTrafficController(clusterID, stop, &GlobalTrafficHandler{RemoteRegistry: r, ClusterID: clusterID}, clientConfig, resyncPeriod)
+	log.Infof("starting service controller clusterID: %v", clusterID)
+	rc.ServiceController, err = admiral.NewServiceController(clusterID, stop, &ServiceHandler{RemoteRegistry: r, ClusterID: clusterID}, clientConfig, 0)
 
 	if err != nil {
-		return fmt.Errorf(" Error with GlobalTrafficController controller init: %v", err)
+		return fmt.Errorf(" Error with ServiceController controller init: %v", err)
```
Comment (nirvanagit) [inline admiral/pkg/clusters/registry.go:127]: nitpick: space before `Error`

File: admiral/pkg/clusters/registry.go [hunk 2]
```diff
@@ -120,31 +120,21 @@ func (r *RemoteRegistry) createCacheController(clientConfig *rest.Config, cluste
...
 
-	log.Infof("starting deployment controller clusterID: %v", clusterID)
-	rc.DeploymentController, err = admiral.NewDeploymentController(clusterID, stop, &DeploymentHandler{RemoteRegistry: r, ClusterID: clusterID}, clientConfig, resyncPeriod)
+	log.Infof("starting global traffic policy controller custerID: %v", clusterID)
+
+	rc.GlobalTraffic, err = admiral.NewGlobalTrafficController(clusterID, stop, &GlobalTrafficHandler{RemoteRegistry: r, ClusterID: clusterID}, clientConfig, 0)
 
 	if err != nil {
-		return fmt.Errorf(" Error with DeploymentController controller init: %v", err)
+		return fmt.Errorf(" Error with GlobalTrafficController controller init: %v", err)
```
Comment (nirvanagit) [inline admiral/pkg/clusters/registry.go:135]: nitpick: space before `Error`

File: admiral/pkg/clusters/types.go [hunk 3]
```diff
@@ -145,6 +146,47 @@ type ServiceHandler struct {
 	ClusterID      string
 }
 
+func (sh *ServiceHandler) Added(obj *k8sV1.Service) {
+	log.Infof(LogFormat, "Added", "service", obj.Name, sh.ClusterID, "received")
+	err := HandleEventForService(obj, sh.RemoteRegistry, sh.ClusterID)
+	if err != nil {
+		log.Infof(err.Error())
```
Comment (nirvanagit) [inline admiral/pkg/clusters/types.go:153]: Should we use Warnf or Errorf instead?  (same comment for other related methods)

File: admiral/pkg/clusters/types.go [hunk 4]
```diff
@@ -145,6 +146,47 @@ type ServiceHandler struct {
...
+func (sh *ServiceHandler) Deleted(obj *k8sV1.Service) {
+	log.Infof(LogFormat, "Deleted", "service", obj.Name, sh.ClusterID, "received")
+	err := HandleEventForService(obj, sh.RemoteRegistry, sh.ClusterID)
+	if err != nil {
+		log.Infof(err.Error())
+	}
+}
+
+func HandleEventForService(svc *k8sV1.Service, remoteRegistry *RemoteRegistry, clusterName string) error {
+	matchingDeployements := remoteRegistry.RemoteControllers[clusterName].DeploymentController.GetDeploymentBySelectorInNamespace(svc.Spec.Selector, svc.Namespace)
```
Comment (nirvanagit) [inline admiral/pkg/clusters/types.go:174]: 1. We could check for nil objects for remoteRegistry
2. Could do the same nil object check for `svc.Spec.Selector`, as there might be services without selectors.

File: admiral/pkg/clusters/types.go [hunk 5]
```diff
@@ -145,6 +146,47 @@ type ServiceHandler struct {
...
+	for _, deployment := range matchingDeployements {
+		HandleEventForDeployment(admiral.Update, &deployment, remoteRegistry, clusterName)
+	}
+	if common.GetAdmiralParams().ArgoRolloutsEnabled {
+		rollouts := remoteRegistry.RemoteControllers[clusterName].RolloutController.GetRolloutBySelectorInNamespace(svc.Spec.Selector, svc.Namespace)
```
Comment (nirvanagit) [inline admiral/pkg/clusters/types.go:179]: We could check for nil objects for remoteRegistry

File: admiral/pkg/controller/admiral/controller.go [hunk 6]
```diff
@@ -102,7 +102,8 @@ func (c *Controller) Run(stopCh <-chan struct{}) {
 	}
 
 	log.Infof("Informer caches synced for controller=%v, current keys=%v", c.name, c.informer.GetStore().ListKeys())
-	wait.Until(c.runWorker, 5*time.Second, stopCh)
+	//wait for 30 seconds for the first time (for all caches to sync)
+	wait.Until(c.runWorker, 30 * time.Second, stopCh)
```
Comment (nirvanagit) [inline admiral/pkg/controller/admiral/controller.go:106]: Curious to know the reason to increase the sync time. I recall, there is a waitForSync or similarly named function given by k8s-api's client-go. Is it possible to use that for individual caches?

File: admiral/pkg/controller/admiral/deployment.go [hunk 7]
```diff
@@ -226,5 +188,13 @@ func (d *DeploymentController) GetDeploymentByLabel(labelValue string, namespace
 		return []k8sAppsV1.Deployment{}
 	}
 
-	return matchedDeployments.Items
+	var filteredDeployments = make([]k8sAppsV1.Deployment, 0)
+
+	for _, deployment := range matchedDeployments.Items {
+		if reflect.DeepEqual(deployment.Spec.Selector.MatchLabels, serviceSelector) {
```
Comment (nirvanagit) [inline admiral/pkg/controller/admiral/deployment.go:194]: We could check for nil objects in deployment.Spec.Selector.MatchLabels

File: admiral/pkg/controller/admiral/rollouts.go [hunk 8]
```diff
@@ -194,20 +194,39 @@ func (roc *RolloutController) Deleted(ojb interface{}) {
...
-		return []argo.Rollout{}
+		return filteredRollouts
+	}
+
+	for _, rollout := range matchedRollouts.Items {
+		match := true
+		for lkey, lvalue := range serviceSelector {
+			// Rollouts controller adds a dynamic label with name rollouts-pod-template-hash to both active and passive replicasets.
+			// This dynamic label is not available on the rollout template. Hence ignoring the label with name rollouts-pod-template-hash
+			if lkey == common.ROLLOUT_POD_HASH_LABEL {
```
Comment (nirvanagit) [inline admiral/pkg/controller/admiral/rollouts.go:217]: nitpick: `ROLLOUT_POD_HASH_LABEL` -> `RolloutPodHashLabel`

File: admiral/pkg/controller/admiral/rollouts.go [hunk 9]
```diff
@@ -194,20 +194,39 @@ func (roc *RolloutController) Deleted(ojb interface{}) {
...
+				continue
+			}
+			value, ok := rollout.Spec.Selector.MatchLabels[lkey]
```
Comment (nirvanagit) [inline admiral/pkg/controller/admiral/rollouts.go:220]: We could check for nil object.

File: admiral/pkg/controller/admiral/service_test.go [hunk 10]
```diff
@@ -75,22 +75,22 @@ func TestServiceCache_Put(t *testing.T) {
 	if serviceCache.getKey(service) != "ns" {
 		t.Errorf("Incorrect key. Got %v, expected ns", serviceCache.getKey(service))
 	}
-	if !cmp.Equal(serviceCache.Get("ns").Service["ns"][service.Name], service) {
-		t.Errorf("Incorrect service fount. Diff: %v", cmp.Diff(serviceCache.Get("ns").Service["ns"], service))
+	if !cmp.Equal(serviceCache.Get("ns")[0], service) {
+		t.Errorf("Incorrect service fount. Diff: %v", cmp.Diff(serviceCache.Get("ns")[0], service))
```
Comment (nirvanagit) [inline admiral/pkg/controller/admiral/service_test.go:79]: typo: `fount` -> `found`

File: admiral/pkg/controller/admiral/controller.go (hunk 6 above)
Comment (aattuluri) [reply admiral/pkg/controller/admiral/controller.go:106]: The controller is already using `waitForCacheSync` before (lines 99-101) in this same file. This delay is an attempt to allow all controllers to complete sync given not all are initialized at the same time.

File: admiral/pkg/clusters/handler_test.go [hunk 11]
```diff
@@ -693,12 +693,11 @@ func TestGetServiceForRolloutCanary(t *testing.T) {
 	selectorMap["app"] = "test"
 
 	service := &coreV1.Service{
+		ObjectMeta: v12.ObjectMeta{Name: SERVICENAME, Namespace: NAMESPACE, CreationTimestamp: v12.NewTime(time.Now())},
```
Comment (nirvanagit) [inline admiral/pkg/clusters/handler_test.go:696]: nitpick: ServiceName, Namespace

File: admiral/pkg/clusters/handler_test.go [hunk 12]
```diff
@@ -711,15 +710,15 @@ func TestGetServiceForRolloutCanary(t *testing.T) {
 	service.Spec.Ports = ports
 
 	stableService := &coreV1.Service{
-		ObjectMeta: v12.ObjectMeta{Name: STABLESERVICENAME, Namespace: NAMESPACE},
+		ObjectMeta: v12.ObjectMeta{Name: STABLESERVICENAME, Namespace: NAMESPACE, CreationTimestamp: v12.NewTime(time.Now().Add(time.Duration(-15)))},
```
Comment (nirvanagit) [inline admiral/pkg/clusters/handler_test.go:713]: nitpick: StableServiceName

File: admiral/pkg/clusters/handler_test.go [hunk 13]
```diff
@@ -711,15 +710,15 @@ func TestGetServiceForRolloutCanary(t *testing.T) {
...
 		Spec: coreV1.ServiceSpec{
 			Selector: selectorMap,
 			Ports:    ports,
 		},
 	}
 
 	canaryService := &coreV1.Service{
-		ObjectMeta: v12.ObjectMeta{Name: CANARYSERVICENAME, Namespace: NAMESPACE},
+		ObjectMeta: v12.ObjectMeta{Name: CANARYSERVICENAME, Namespace: NAMESPACE, CreationTimestamp: v12.NewTime(time.Now().Add(time.Duration(-15)))},
```
Comment (nirvanagit) [inline admiral/pkg/clusters/handler_test.go:721]: Nitpick: CanaryServiceName

File: admiral/pkg/controller/admiral/deployment.go [hunk 14]
```diff
@@ -226,5 +188,16 @@ func (d *DeploymentController) GetDeploymentByLabel(labelValue string, namespace
 		return []k8sAppsV1.Deployment{}
 	}
 
-	return matchedDeployments.Items
+	var filteredDeployments = make([]k8sAppsV1.Deployment, 0)
+
+	for _, deployment := range matchedDeployments.Items {
+		if deployment.Spec.Selector == nil || deployment.Spec.Selector.MatchLabels == nil {
```
Comment (nirvanagit) [inline admiral/pkg/controller/admiral/deployment.go:194]: We could check for nil values for `.Selector` and `.MatchLabels`

File: admiral/pkg/controller/admiral/deployment.go [hunk 15]
```diff
@@ -211,10 +173,10 @@ func (d *DeploymentController) shouldIgnoreBasedOnLabels(deployment *k8sAppsV1.D
...
 }
 
-func (d *DeploymentController) GetDeploymentByLabel(labelValue string, namespace string) []k8sAppsV1.Deployment {
-	matchLabel := common.GetGlobalTrafficDeploymentLabel()
+func (d *DeploymentController) GetDeploymentBySelectorInNamespace(serviceSelector map[string]string, namespace string) []k8sAppsV1.Deployment {
+
 	labelOptions := meta_v1.ListOptions{}
-	labelOptions.LabelSelector = fmt.Sprintf("%s=%s", matchLabel, labelValue)
+
 	matchedDeployments, err := d.K8sClient.AppsV1().Deployments(namespace).List(labelOptions)
```
Comment (nirvanagit) [inline admiral/pkg/controller/admiral/deployment.go:180]: We should be able to use listOptions here to get the matching deployments.

Comment (aattuluri) [reply admiral/pkg/controller/admiral/deployment.go:180]: Not really as we want to match on the selector and not the labels of the deployment itself - https://stackoverflow.com/questions/69903346/list-deployments-using-apimachinery-by-the-spec-selector-matchlabels-key

File: admiral/pkg/controller/admiral/rollouts.go [hunk 16]
```diff
@@ -194,20 +194,43 @@ func (roc *RolloutController) Deleted(ojb interface{}) {
...
+			continue
+		}
+		match := true
+		for lkey, lvalue := range serviceSelector {
+			// Rollouts controller adds a dynamic label with name rollouts-pod-template-hash to both active and passive replicasets.
+			// This dynamic label is not available on the rollout template. Hence ignoring the label with name rollouts-pod-template-hash
+			if lkey == common.RolloutPodHashLabel {
+				continue
+			}
+			value, ok := rollout.Spec.Selector.MatchLabels[lkey]
```
Comment (nirvanagit) [inline admiral/pkg/controller/admiral/rollouts.go:224]: We can use the labelSelector parameter in the list option to find deployments matching given labels.

File: admiral/pkg/controller/common/common.go [hunk 17]
```diff
@@ -26,15 +26,16 @@ const (
...
+	MulticlusterIngressGateway           = "istio-multicluster-ingressgateway"
+	LocalAddressPrefix                   = "240.0"
+	NodeRegionLabel                      = "failure-domain.beta.kubernetes.io/region"
+	SpiffePrefix                         = "spiffe://"
+	SidecarEnabledPorts                  = "traffic.sidecar.istio.io/includeInboundPorts"
+	Default                              = "default"
+	AdmiralIgnoreAnnotation              = "admiral.io/ignore"
+	AdmiralCnameCaseSensitive            = "admiral.io/cname-case-sensitive"
+	BlueGreenRolloutPreviewPrefix        = "preview"
+	RolloutPodHashLabel           string = "rollouts-pod-template-hash"
```
Comment (nirvanagit) [inline admiral/pkg/controller/common/common.go:38]: I think the formatting is off starting line 29.

File: admiral/pkg/clusters/types.go [hunk 18]
```diff
@@ -243,7 +290,6 @@ func HandleEventForRollout(event admiral.EventType, obj *argo.Rollout, remoteReg
 
 // helper function to handle add and delete for DeploymentHandler
 func HandleEventForDeployment(event admiral.EventType, obj *k8sAppsV1.Deployment, remoteRegistry *RemoteRegistry, clusterName string) {
-	log.Infof(LogFormat, event, "deployment", obj.Name, clusterName, "Received")
```
Comment (nirvanagit) [inline admiral/pkg/clusters/types.go:246]: Do we not need this log anymore?

Comment (aattuluri) [reply admiral/pkg/clusters/types.go:246]: This has been pushed further up as this method is being invoked for other events.

**Review by nirvanagit — APPROVED** [review]

**Pull request merged**